File: google/main/src/core/engine/SearchEngineApp.java
package main.src.core.engine;

import main.src.core.structures.Document;
import main.src.utils.FileHandler;
import main.src.utils.Prettifier;
import main.src.core.query.Query;

import java.util.*;
import java.lang.IllegalArgumentException;


public class SearchEngineApp {

    private Scanner scanner = new Scanner(System.in);
    private String resourcesDirectory;
    private InvertedIndex index;

    public SearchEngineApp(String resourcesDirectory){
        this.resourcesDirectory = resourcesDirectory;
        System.out.print("Indexing...");
        this.index = new InvertedIndex(this.resourcesDirectory);
        System.out.println("DONE!\n");
        System.out.println("Welcome to google! Copyright(c) 2020 v1.0.1");
        help();
    }

    private boolean handleCommand(String command, String arguments){
        if (command.equals("search"))    search(arguments);
        else if (command.equals("view")) view(arguments);
        else if (command.equals("help")) help();
        else if (command.equals("exit")) return false;
        else System.out.println(String.format("\'%s\' is not recognized as an internal or external command."));
        return true;
    }

    public void handleCommands() {
        String input, command, arguments;
        int firstSpaceIndex;
        do {
            System.out.print("google> ");
            input = scanner.nextLine();
            firstSpaceIndex = input.indexOf(' ');
            command = input.substring(0, firstSpaceIndex);
            arguments = input.substring(firstSpaceIndex + 1);
        } while(handleCommand(command, arguments));
    }

    private void search(String arguments) {
        try {
            HashSet<Document> results = Engine.getQueryResults(new Query(arguments), index);
            System.out.println(Prettifier.prettify(results));
        } catch (IllegalArgumentException e) {
            System.out.println(e.getMessage());
        }
    }

    private void view(String arguments) {
        System.out.println(FileHandler.loadFile(this.resourcesDirectory + "/" + arguments));
    }

    private void help() {
        System.out.println("\nsearch <terms>\n- search for terms\n term : AND feature\n+term : OR  feature\n-term : EXC feature");
        System.out.println("\nhelp\n- display commands\n");
        System.out.println("view <docId>\n- view the documents\n");
    }
}

File: Bagherzadeh/BagherEngine/App/SearchEngineApp.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

using BagherEngine.Elastic;
using BagherEngine.Exceptions;
using BagherEngine.Models;
using BagherEngine.QuerySystem;
using BagherEngine.Utils;

namespace BagherEngine.App
{
	public class AppMessages
	{
		public const string NoResultsFoundMessage = "No results found!";
		public const string NoKeyWordsPassedMessage = "No keywords passed!";
		public const string NoDocumentIdPassedMessage = "No documentId passed!";
		public const string NoPathPassedMessage = "No path passed!";
		public const string DirectoryNotFoundMessage = "Directory not found!";
		public const string FileNotFoundMessage = "File not found!";
		public const string HelpMessage = "search <terms> -- view <documentId> -- help -- import <pathToDocuments>";
	}

	public class SearchEngineApp : ConsoleApp
	{
		private const string AppName = "Bagherzadeh";
		private const string Year = "2020";
		private const string Version = "v1.0.1";
		private const string IndexName = "bagher-documents";

		private readonly string ResourcesDirectory;

		private Engine bagherEngine = Engine.GetInstance();

		public override void Intro()
		{
			WriteLine($"Welcome to {AppName}! Copyright(c) Team8 {Year} {Version}");
		}

		public override void Help()
		{
			WriteLine(AppMessages.HelpMessage);
		}

		public SearchEngineApp(string resourcesDirectory) : base()
		{
			ResourcesDirectory = resourcesDirectory;
			prompt = AppName + "> ";
		}

		public override bool HandleCommand()
		{
			switch (command)
			{
				case "search":
					Search(arguments);
					break;
				case "view":
					View(arguments);
					break;
				case "help":
					Help();
					break;
				case "import":
					Import(arguments);
					break;
				case "exit":
					return false;
				default:
					WriteLine($"'{command}' is not recognized as an internal or external command.");
					break;
			}

			return true;
		}

		private string GetSearchResults(string arguments)
		{
			var results = bagherEngine.GetQueryResults(new Query(arguments));
			if (!results.Any())
				return AppMessages.NoResultsFoundMessage;
			return Prettifier<Document>.Prettify(results);
		}

		public void Search(string arguments)
		{
			if (String.IsNullOrEmpty(arguments))
			{
				WriteError(AppMessages.NoKeyWordsPassedMessage);
				return;
			}
			try
			{
				var results = GetSearchResults(arguments);
				WriteLine(results);
			}
			catch (Exception e)
			{
				WriteError(e.Message);
			}
		}

		public void View(string arguments)
		{
			if (String.IsNullOrEmpty(arguments))
			{
				WriteError(AppMessages.NoDocumentIdPassedMessage);
				return;
			}
			try
			{
				WriteLine(FileHandler.GetFileContent(Path.Combine(ResourcesDirectory, arguments)));
			}
			catch (FileNotFoundException)
			{
				WriteError(AppMessages.FileNotFoundMessage);
			}
		}

		public void Import(string arguments)
		{
			if (String.IsNullOrEmpty(arguments))
			{
				WriteError(AppMessages.NoPathPassedMessage);
				return;
			}
			Console.WriteLine($"\nImporting documents from {arguments}...");
			try
			{
				var documents = FileHandler.GetDocumentsFromFolder(arguments);
				var ids = documents.Select(doc => doc.Id).ToList();
				var importer = new Importer<Document>();
				importer.Import(documents, IndexName, ids);
				Console.WriteLine($"[DONE]\n");
			}
			catch (DirectoryNotFoundException)
			{
				WriteError(AppMessages.DirectoryNotFoundMessage);
			}
		}

		private void WriteLine(string output)
		{
			Console.WriteLine("\n" + output + "\n");
		}

		private void WriteError(string error)
		{
			Console.ForegroundColor = ConsoleColor.Red;
			WriteLine(error);
			Console.ResetColor();
		}
	}
}
